src/lib.rs

```diff
@@ -46,15 +46,22 @@ mod util;
 
 pub use crate::bit_machine::exec;
 pub use crate::core::{CommitNode, RedeemNode};
+use crate::inference::UnificationArrow;
 use std::fmt;
 
 /// Error type for simplicity
 #[derive(Debug)]
 pub enum Error {
     /// A type cannot be unified with another type
-    TypeCheck(&'static str),
+    Unification(&'static str),
     /// A type is recursive (i.e., occurs within itself), violating the "occurs check"
     OccursCheck,
+    /// A DAG cannot be created because the children of the root cannot be unified
+    TypeCheck(
+        [&'static str; 2],
+        Option<UnificationArrow>,
+        Option<UnificationArrow>,
+    ),
```

Comment:
In 5f4ba25c37dc8e04c43c0b2fb3d412c3922511e1:

These should all be named fields, and the two strings should not be bundled into one array.